keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,575 @@
+apiserver.
+
+This is kind of an odd case, but [it has been seen in the wild], in
+the case of a dual-stack IPv6-primary cluster that contains some nodes
+that are on an IPv4-only network, which can therefore only host
+IPv4-only pods.
+
+[it has been seen in the wild]: https://github.com/kubernetes/enhancements/issues/563#issuecomment-814560270
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+In theory, an administrator might be running an apiserver on a
+dual-stack host but not want it to be exposed as a dual-stack service.
+However, like most golang-based servers, the apiserver _listens_ on
+both IP families by default unless explicitly configured to only
+listen on a single IP, so the administrator in this case would already
+be losing, and just not realizing it. Making the apiserver advertise
+its secondary IP address in this case would arguably be a security
+_improvement_, since it would make it easier for the administrator to
+notice that it was not configured how they had wanted.
+
+## Design Details
+
+### kube-apiserver Command-Line Arguments
+
+FIXME
+
+### Pod environment / client-go
+
+To fully support the "wrong-single-stack pod" use case, we need to
+make `rest.InClusterConfig()` use a dual-stack configuration.
+Currently it tries to connect to `KUBERNETES_SERVICE_HOST`.
+
+Changing anything about `KUBERNETES_SERVICE_HOST` would likely break
+some existing users, and should be avoided.
+
+Although we could pass a `KUBERNETES_SERVICE_HOST_SECONDARY` variable
+to the pod, `rest.InClusterConfig()` cannot make use of two separate
+IPs; it must generate a configuration containing a single apiserver
+URL.
+
+Another possibility would be for it to use
+`https://kubernetes.default.svc.cluster.local` rather than
+`https://${KUBERNETES_SERVICE_HOST}`. Presumably though, doing this
+unconditionally would break some clusters, so this is also not an
+option.
+
+To make this work, `rest.InClusterConfig()` would have to introspect
+its own local network setup (eg, using `net.Interfaces()`) to see if
+it would be able to reach `KUBERNETES_SERVICE_HOST`, and then use an
+alternative (`KUBERNETES_SERVICE_HOST_SECONDARY` or
+`kubernetes.default.svc.cluster.local`) only if the primary apiserver
+IP is unreachable from that pod.
+
+There was [discussion in an earlier KEP] about whether it is actually
+safe to add new `KUBERNETES_*` environment variables, given that we
+don't restrict users from using variables with such names for their
+own purposes...)
+
+It's possible that there is no solution that is nice enough that it's
+actually worth trying to make it work for the admittedly-edge-case-y
+scenario of wrong-single-stack pods...
```

Comment:
this ... also, there is a workardound that is using `kubernetes.default` as `KUBERNETES_HOST`

Reply:
that you already cover below too :) ... useless comment